### ISSUE TITLE
`FakeFS` and `CSV` issues in Ruby v2.6.x

### ISSUE THREAD

**Issue by cbortz** [fakefs/fakefs#428]
Hey y'all,

Thanks for the great library!

---

Consider the following code:

```ruby
absolute_path_to_local_csv = "/path/to/file.csv"

FakeFS do
  FakeFS::FileSystem.clone(absolute_path_to_local_csv)
  CSV.read(absolute_path_to_local_csv)
end
```

In Ruby v2.3.x - v2.5.x, a nested array is returned such that each value is an array of columns values representing each row in the CSV (expected behavior).

However, after upgrading to Ruby v2.6.0, the same code produces this error:

```bash
undefined method `string' for #<FakeFS::File:0x00007fdb072b73e0>
```

I've tested this with Ruby v2.6.0 and v2.6.1, using multiple versions of `FakeFS`, from v0.11.0 to v0.19.1.

As far as I can tell, both `CSV.open` and `CSV.foreach` are effected by this, but I would assume `CSV.open` as well as other class methods may result in the same error.

---

I noticed that this may be related to a much older issue #192 (closed). Please let me know if I can provide any additional information to help debug this!

**Comment by grosser** [comment]
sounds like we need a new 'string' method ... PR welcome ... ideally with a test-case with CSV.open :)

**Comment by cbortz** [comment]
> PR welcome ... ideally with a test-case with CSV.open :)

As I started to work on this, I found that when starting a completely pristine project with Ruby v2.6.1 I actually don't encounter this issue. What's odd is when upgrading my Rails application from 2.5.3 to 2.6.1, the _only_ error I encountered was this one.

At this point, it seems more likely that it's an environment/dependency issue rather than a `fakefs` issue.

Thanks for responding the the issue. I'll close it for now, and re-open it if this ends up actually being relevant to the project.

Cheers!

**Comment by cbortz** [comment]
Alright, so I kept looking into it, and I've discovered some more (hopefully useful) information. Ultimately I can get reproduce this error with `CSV.read` but not `CSV.open` That's because the `CSV.open` doesn't invoke the method that results in the error, whereas `CSV.read` does.

This explanation is a bit lengthly. I'll put a tl;dr in there towards the end. 👍 

---

In the tests I was running, I used a very small CSV, for example:

```
a,b,c
1,2,3
```

In my test, I simply ran this:

```ruby
path = "#{`pwd`.chomp}/file.csv"

FakeFS do
  FakeFS::FileSystem.clone(path)
  CSV.read(path)
end
```

No exception was raised. I'd accepted there wasn't actually a bug and closed this ticket. However, I found that when I increase the file size, the `#read` gets chunked and that's where the exception happens. Here's the example CSV:

```
headercolumn_1,headercolumn_2,headercolumn_3,headercolumn_4,headercolumn_5,headercolumn_6,headercolumn_7
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x,row_x_column_x
```

Recently `CSV` went through refactor where some [parsing logic was extracted](https://github.com/ruby/csv/commit/d49380a4a1a6ca1626258d9dd3fdd9860367bf8a). Here's how I understand it:

`CSV.read` will first open the CSV, and [then invoke `#read`](https://github.com/ruby/csv/blob/master/lib/csv.rb#L714). Then `#read` invokes the Enumerable method `#to_a`, which relies on `CSV#each`. `CSV#each` then calls `CSV#parse`, which is an instance of `CSV::Parser`.

If you follow the trail of instantiating a `::Parser`, you'll find that `#resolve_row_separator` is invoked, which shovels a value into `@samples` by "sampling" the data via chunking. Here's that code:

```ruby
if @input.is_a?(File)
  chunk_size = 32 * 1024
else
  chunk_size = 1024
end
```

At this point, we'll end up with a chunk_size of 1024 because `FakeFS::File#is_a?(File)` will not be true. This seems innocent enough out of context, but later this happens:

[`CSV::Parser#parse` invokes `CSV::Parser#build_scanner `](https://github.com/ruby/csv/blob/master/lib/csv/parser.rb#L245), which under the default conditions (ENV `CSV_PARSER_SCANNER_TEST` is not set to `"yes"`) will be defined as such (comments added pointing out key behavior points):

```ruby
def build_scanner
  string = nil
  if @samples.empty? and @input.is_a?(StringIO)
    string = @input.string
  elsif @samples.size == 1 and @input.respond_to?(:eof?) and @input.eof?

    # (1) when chunk_size <= 1024, EOF is reached and `string` is now a `String` (see (2) below)

    string = @samples[0]
  end
  if string
    unless string.valid_encoding?
      message = "Invalid byte sequence in #{@encoding}"
      raise MalformedCSVError.new(message, @lineno + 1)
    end

    # (2) Since string is a `String`, and EOF reached, return here

    Scanner.new(string)
  else

    # (3) Since `string` is still `nil` for chunks larger than 1024, this is invoked

    inputs = @samples.collect do |sample|

      # (4) Each sample (`<String>`) is mapped to a StringIO object, not a problem

      StringIO.new(sample)
    end

    # (5) But then the original @input (`<FakeFS::File>`) is shoveled in

    inputs << @input

    # (6) The inputs may look like this: `[<StringIO>, <FakeFS::File>]`
    # which will cause a problem in `InputScanner`

    InputsScanner.new(inputs, @encoding)
  end
end
```

## TL;DR

So, finally, we get to the `InputsScanner`. Again, assuming we have a CSV file with a total chunk_size greater than 1024, we will pass `inputs` when instantiating an InputScanner that look like:

```ruby
InputScanner.new([<StringIO>, <FakeFS::File>], encoding)
```

Ok, so on to where the error actually happens. `InputsScanner#initialize` invokes [`#read_chunks`](https://github.com/ruby/csv/blob/master/lib/csv/parser.rb#L138), which includes this block of code:

```ruby
input = @inputs.first
case input
when StringIO
  string = input.string
  raise InvalidEncoding unless string.valid_encoding?
  @scanner = StringScanner.new(string)
  @inputs.shift
  @last_scanner = @inputs.empty?
  true
else
  # ...
end
```

As the scanner iterates over the inputs,  eventually we get to the `FakeFS::File` object, which returns `true` for `#is_a?(StringIO)`, as applied in https://github.com/fakefs/fakefs/commit/ec6304766b7372343d8f611ac8e488aa2128238d, which was interestingly listed as a fix many years ago for a similar exception #192. How times have changed, right? 😄 

---

I hope that is enough information to highlight the issue.

@grosser I'm still happy to take a stab at fixing this. You'd originally requested the test to be written against `CSV.open`, but given the actual behavior, I believe this should be tested against `CSV.read`. Thoughts?

**Comment by grosser** [comment]
CSV.read sounds right!
